neurox/data/filter.py

```diff
@@ -0,0 +1,162 @@
+    concept_filter: a set of words or a regex object or a function
+    num_layers: number of layers
+    
+    Returns
+    -------
+    Saves a word file, a binay label file and their activations
+    
+    Example
+    -------
+    create_concept_data(tokens, activations, re.compile(r'^\w\w$'), 12) select words of two characters only as a positive class
+    create_concept_data(tokens, activations, {'is', 'can'}, 12) select occrrences of 'is' and 'can' as a positive class
+    """
+    
+    filter_fn = None
+    if isinstance(concept_filter, set):
+        filter_fn = lambda x: x in concept_filter
+    elif isinstance(concept_filter, re.Pattern):
+        filter_fn = lambda x: concept_filter.match(x)
+    elif isinstance(concept_filter, collections.Callable):
+        filter_fn = concept_filter
+    else:
+        print ("ERROR: doest not belong to any configuration")
+        return
+        
+    positive_class_words = []
+    positive_class_activations = []
+    negative_class_words = []    
+    negative_class_activations = []
+    
+    print ("Creating concept dataset ...")
+    for sidx, sentences in enumerate(tokens['source']):
+        for widx, word in enumerate(sentences):
+            if filter_fn(word):
+                positive_class_words.append(word)
+                positive_class_activations.append(activations[sidx][widx].reshape((num_layers, 1, -1)))
+            else:
+                negative_class_words.append(word)
+                negative_class_activations.append(activations[sidx][widx].reshape((num_layers, 1, -1)))
+
+    negative_class_words, negative_class_activations = _balance_negative_class(negative_class_words, negative_class_activations, len(positive_class_words))    
+    
+    print ("Number of Positive examples: ", len(positive_class_words))
+    
+    labels = (['positive']*len(positive_class_words)) + ['negative']*len(positive_class_words)
+    
+    save_files(positive_class_words+negative_class_words, labels, positive_class_activations+negative_class_activations, outputfile)
+    
+    
+def _balance_negative_class(words, activations, positive_class_size):
+    """
+    Helper function to shuffle the negative class instances and select the number of instances equal to 
+    the positive class
+    
+    Parameters
+    ----------
+    words: list
+        A list of words
+    activations: list
+        A list of word-wise activations 
+    positive_class_size: number of words to select
+    
+    Returns
+    -------
+    A list of words and activations equal in size to the passed length
+    
+    """    
+    print ("Balancing Negative class ...")
+    indices = list(range((len(words))))
+    np.random.shuffle(indices)
+
+    swords = []
+    sactivations = []
+    for i in range(positive_class_size):
+        swords.append(words[indices[i]])
+        sactivations.append(activations[indices[i]])
+    
+    return swords, sactivations
+    
+def save_files(words, labels, activations, outputfile):
```

Comment:
Save file should now use `ActivationWriter`. We can do this together or you can see how its done in transformers_extractor.